File: Task/Mayan-calendar/Nim/mayan-calendar.nim
import math, strformat, times

const

  Tzolk´in = ["Imix´", "Ik´", "Ak´bal", "K´an", "Chikchan", "Kimi", "Manik´", "Lamat", "Muluk", "Ok",
              "Chuwen", "Eb", "Ben", "Hix", "Men", "K´ib´", "Kaban", "Etz´nab´", "Kawak", "Ajaw"]

  Haab´ = ["Pop", "Wo´", "Sip", "Sotz´", "Sek", "Xul", "Yaxk´in", "Mol", "Ch´en", "Yax",
           "Sak´", "Keh", "Mak", "K´ank´in", "Muwan", "Pax", "K´ayab", "Kumk´u", "Wayeb´"]

let
  creationTzolk´in = initDateTime(21, mDec, 2012, 0, 0, 0, utc())
  zeroHaab´ = initDateTime(2, mApr, 2019, 0, 0, 0, utc())

func daysPerMayanMonth(month: string): int =
  if month == "Wayeb´": 5 else: 20

proc tzolkin(gregorian: DateTime): string =
  let deltaDays = (gregorian - creationTzolk´in).inDays
  let rem = floorMod(deltaDays, 13)
  result = $(if rem <= 9: rem + 4 else: rem - 9) & ' ' & Tzolk´in[floorMod(deltaDays - 1, 20)]

proc haab(gregorian: DateTime): string =
  let rem = floorMod((gregorian - zeroHaab´).inDays, 365)
  let month = Haab´[(rem + 1) div 20]
  let dayOfMonth = rem mod 20 + 1
  result = if dayOfMonth < daysPerMayanMonth(month): &"{dayOfMonth} {month}" else: &"Chum {month}"

proc toLongDate(gregorian: DateTime): string =
  var delta = (gregorian - creationTzolk´in).inDays + 13 * 360 * 400
  var baktun = delta div (360 * 400)
  delta = delta mod (400 * 360)
  let katun = delta div (20 * 360)
  delta = delta mod (20 * 360)
  let tun = delta div 360
  delta = delta mod 360
  let winal = delta div 20
  let kin = delta mod 20
  result = &"{baktun:2} {katun:2} {tun:2} {winal:2} {kin:2}"

proc nightLord(gregorian: DateTime): string =
  var rem = (gregorian - creationTzolk´in).inDays mod 9
  if rem <= 0: rem += 9
  result = 'G' & $rem


when isMainModule:

  let testDates = [initDateTime(21, mNov, 1963, 0, 0, 0, utc()),
                   initDateTime(19, mJun, 2004, 0, 0, 0, utc()),
                   initDateTime(18, mDec, 2012, 0, 0, 0, utc()),
                   initDateTime(21, mDec, 2012, 0, 0, 0, utc()),
                   initDateTime(19, mJan, 2019, 0, 0, 0, utc()),
                   initDateTime(27, mMar, 2019, 0, 0, 0, utc()),
                   initDateTime(29, mFeb, 2020, 0, 0, 0, utc()),
                   initDateTime(01, mMar, 2020, 0, 0, 0, utc()),
                   initDateTime(16, mMay, 2071, 0, 0, 0, utc()),
                   initDateTime(02, mfeb, 2020, 0, 0, 0, utc())]

echo "Gregorian      Long Count       Tzolk´in  Haab´       Nightlord"
echo "———————————————————————————————————————————————————————————————"
for date in testDates:
  let dateStr = date.format("YYYY-MM-dd")
  echo &"{dateStr:14} {date.toLongDate:16} {tzolkin(date):9} {haab(date):14} {nightLord(date)}"

File: Task/Mayan-calendar/Java/mayan-calendar.java
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.List;

public final class MayanCalendar {

	public static void main(String[] aArgs) {
		List<LocalDate> testDates = List.of( LocalDate.parse("2004-06-19"),
											 LocalDate.parse("2012-12-18"),
											 LocalDate.parse("2012-12-21"),
											 LocalDate.parse("2019-01-19"),
											 LocalDate.parse("2019-03-27"),
											 LocalDate.parse("2020-02-29"),
											 LocalDate.parse("2020-03-01"),
											 LocalDate.parse("2071-05-16"),
											 LocalDate.parse("2020-02-02") );
		
		System.out.println("Gregorian      Long Count         Tzolk'in    Haab'         Lords of the Night");
		System.out.println("------------------------------------------------------------------------------");
		for ( LocalDate date : testDates ) {
			System.out.println(String.format("%-15s%-19s%-12s%-18s%s",
				date.toString(), longCount(date), tzolkin(date), haab(date), lordsOfTheNight(date)));
		}
	}
	
	private static String lordsOfTheNight(LocalDate aGregorian) {
		long daysBetween = ChronoUnit.DAYS.between(CREATION_TZOLKIN, aGregorian);
		long remainder = Math.floorMod(daysBetween, 9);		
		return "G" + ( ( remainder <= 0 ) ? remainder + 9 : remainder );
	}
	
	private static String longCount(LocalDate aGregorian) {
		long daysBetween = ChronoUnit.DAYS.between(CREATION_TZOLKIN, aGregorian) + 13 * 360 * 400;
		long baktun = Math.floorDiv(daysBetween, 360 * 400);	
		daysBetween = Math.floorMod(daysBetween, 360 * 400);	
		long katun = Math.floorDiv(daysBetween, 20 * 360);	
		daysBetween = Math.floorMod(daysBetween, 20 * 360);
		long tun = Math.floorDiv(daysBetween, 360);	
		daysBetween = Math.floorMod(daysBetween, 360);	
		long winal = Math.floorDiv(daysBetween, 20);	
		long kin = Math.floorMod(daysBetween, 20);
		
		StringBuilder result = new StringBuilder();
		for ( long number : List.of( baktun, katun, tun, winal, kin ) ) {
			String value = String.valueOf(number) + ".";
			result.append( number <= 9 ? "0" + value : value );
		}
		return result.toString().substring(0, result.length() - 1);
	}
	
	private static String haab(LocalDate aGregorian) {
		long daysBetween = ChronoUnit.DAYS.between(ZERO_HAAB, aGregorian);
		int remainder = Math.floorMod(daysBetween, 365);
		String month = Haab.get(Math.floorDiv(remainder + 1, 20));
        int dayOfMonth = Math.floorMod(remainder, 20) + 1;
		return ( dayOfMonth < daysPerMayanMonth(month) ) ? dayOfMonth + " " + month : "Chum " + month;
	}

	private static String tzolkin(LocalDate aGregorian) {
		long daysBetween = ChronoUnit.DAYS.between(CREATION_TZOLKIN, aGregorian);
		int remainder = Math.floorMod(daysBetween, 13);	
		remainder += ( remainder <= 9 ) ? 4 : -9;
		return remainder + " " + Tzolkin.get(Math.floorMod(daysBetween - 1, 20));
	}
	
	private static int daysPerMayanMonth(String aMonth) {
		return ( aMonth == "Wayeb'" ) ? 5 : 20;
	}
	
	private static List<String> Tzolkin = List.of( "Imix'", "Ik'", "Ak'bal", "K'an", "Chikchan", "Kimi", "Manik'",
		"Lamat", "Muluk", "Ok", "Chuwen", "Eb", "Ben", "Hix", "Men", "K'ib'", "Kaban", "Etz'nab'", "Kawak", "Ajaw" );

	private static List<String> Haab = List.of( "Pop", "Wo'", "Sip", "Sotz'", "Sek", "Xul", "Yaxk'in", "Mol",
		"Ch'en", "Yax", "Sak'", "Keh", "Mak", "K'ank'in", "Muwan", "Pax", "K'ayab", "Kumk'u", "Wayeb'" );
	
	private static final LocalDate CREATION_TZOLKIN = LocalDate.parse("2012-12-21");
	private static final LocalDate ZERO_HAAB = LocalDate.parse("2019-04-02");

}
